src/Solver.cpp

```diff
@@ -315,16 +391,90 @@ void Solver::setPolarQuadrature(PolarQuad* polar_quad) {
 
 
 /**
- * @brief Sets the threshold for source convergence (>0)
- * @param source_thresh the threshold for source convergence
+ * @brief Sets the threshold for source/flux convergence.
+ * @brief The default threshold for convergence is 1E-5.
+ * @param source_thresh the threshold for source/flux convergence
+ */
+void Solver::setConvergenceThreshold(FP_PRECISION threshold) {
+
+  if (threshold <= 0.0)
+    log_printf(ERROR, "Unable to set the convergence threshold to %f "
+               "since it is not a positive number", threshold);
+
+  _converge_thresh = threshold;
+}
+
+
+/**
+ * @brief Assign a fixed source for a flat source region and energy group.
+ * @details This is a helper routine to perform error checking for the
+ *          subclasses which store the source in the appropriate array.
+ * @param fsr_id the flat source region ID
+ * @param group the energy group
+ * @param source the volume-averaged source in this group
+ */
+void Solver::setFixedSourceByFSR(int fsr_id, int group, FP_PRECISION source) {
+  
+  if (group <= 0 || group > _num_groups)
+    log_printf(ERROR,"Unable to set fixed source for group %d in "
+               "in a %d energy group problem", group, _num_groups);
+
+  if (fsr_id < 0 || fsr_id >= _num_FSRs)
+    log_printf(ERROR,"Unable to set fixed source for FSR %d with only "
+               "%d FSRs in the geometry", fsr_id, _num_FSRs);
+}
+
+
+/**
+ * @brief Assign a fixed source for a Cell and energy group.
+ * @details This routine will add the fixed source to all instances of the
+ *          Cell in the geometry (e.g., all FSRs for this Cell).
+ * @param fsr_id the Cell of interest
+ * @param group the energy group
+ * @param source the volume-averaged source in this group
+ */
+void Solver::setFixedSourceByCell(Cell* cell, int group, FP_PRECISION source) {
+
+  /* Recursively add the source to all Cells within a FILL type Cell */
+  if (cell->getType() == FILL) {
+    std::map<int, Cell*> cells = cell->getAllCells();
+    std::map<int, Cell*>::iterator iter;
+    for (iter = cells.begin(); iter != cells.end(); ++iter)
+      setFixedSourceByCell(iter->second, group, source);
+  }
+
+  /* Aadd the source to all FSRs for this MATERIAL type Cell */
+  else {
+    CellBasic* fsr_cell;
+    
+    for (int r=0; r < _num_FSRs; r++) {
+      fsr_cell = _geometry->findCellContainingFSR(r);
+      if (cell->getId() == fsr_cell->getId())
+        setFixedSourceByFSR(r, group, source);
+    }
+  }
+}
+
```

Comment:
Alright, so it was painful, but I took your idea to heart and overhauled the ``Cell`` implementation in OpenMOC so that this will work now.  You should definitely review the latest code, particularly that in "Cell.h" and "Cell.cpp". 

There are no longer ``CellFill`` and ``CellBasic`` classes, but rather a single ``Cell`` class which can take a ``Material``, ``Universe`` or ``Lattice`` fill. The ``Cell`` class can automatically discretize itself into rings/sectors regardless of its fill type (previously it could only do so if it had a ``Material`` inside). When rings/sectors are created, a new fill ``Universe`` is created and each ring/sector sub-``Cell`` is added to the ``Universe``. 

This new ``Cell`` methodology will allow fixed sources to be assigned to the master/parent ``Cell`` created by the user, and the source will be assigned to each child sub-``Cell`` under-the-hood.